src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -16,96 +16,113 @@ open VectorMatrix
+          (let h = Hashtbl.create @@ IntMap.cardinal !(snd m) in (* #of bindings is a tight upper bound on the number of CCs and thus on the number of different lookups *)
+           fun x -> (* standard memoization wrapper *)
+             try Hashtbl.find h x with Not_found ->
+               let r = bumpvar 0 x offsetlist in
+               Hashtbl.add h x r;
+               r)
       in
-      let add_offset_to_array_entry (var, offs) = (* uses offset_map to obtain a new var_index, that is consistent with the new reference indices *)
-        Option.map (fun var_index -> var_index + offset_map.(var_index)) var, offs in
-      let m' = make_empty_array (length m + length indexes)
-      in Array.iteri (fun j eq -> m'.(j + offset_map.(j)) <- add_offset_to_array_entry eq) m;
-      m'(* produces a consistent new conj. of equalities *)
-
-  let remove_variables_from_domain m indexes =
-    let nr_removed_colums = length indexes in
-    if nr_removed_colums = 0 || length m = 0 then m
-    else
-    if length m = nr_removed_colums then [||] else
-      let offset_map = Array.make (Array.length m) 0
-      (* maps each variable to the number of variables that are removed before this variable *)
-      in let _ = Array.fold_lefti
-             (fun offset index _ ->
-                if offset < nr_removed_colums && indexes.(offset) = index then offset + 1
-                else (offset_map.(index) <- offset; offset))
-             0 offset_map in
-      let remove_offset_from_array_entry (var, offs) =
-        Option.map (fun var_index -> var_index - offset_map.(var_index)) var, offs in
-      Array.(filteri (fun i _ -> not @@ Array.mem i indexes) m (* filter out removed variables*)
-             |> map remove_offset_from_array_entry) (* adjust variable indexes *)
-
-  let remove_variables_from_domain m cols = timing_wrap "del_cols" (remove_variables_from_domain m) cols
-
-  let is_empty m = length m = 0
+      let rec bumpentry k (refvar,offset) = function (* directly bumps lhs-variable during a run through indexes, bumping refvar explicitely with a new lookup in indexes *)
+        | (tbl,delta,head::rest) when k>=head            -> bumpentry k (refvar,offset) (tbl,delta+1,rest) (* rec call even when =, in order to correctly interpret double bumps *)
+        | (tbl,delta,list) (* k<head or list=[] *) -> (IntMap.add (op k delta) (BatOption.map (memobumpvar) refvar, offset) tbl, delta, list)
+      in
+      let (a,_,_) = IntMap.fold (bumpentry) !(snd m) (IntMap.empty,0,offsetlist) in (* Build new map during fold with bumped key/vals *)
+      (op (get_dim m) (Array.length indexes), ref a)
+
+  let modify_variables_in_domain m cols op = let res = modify_variables_in_domain m cols op in if M.tracing then
+      M.tracel "modify_dims" "dimarray bumping with (fun x -> x + %d) at positions [%s] in { %s } -> { %s }"
+        (op 0 1)
+        (Array.fold_right (fun i str -> (string_of_int i) ^ ", " ^ str) cols "")
+        (show @@ !( snd m))
```

Comment:
Are the `@@`s needed here?

Reply:
No, initially a brain fart, and then copy-pasted all over the file